plutus-contract/src/Plutus/Contract/Test/ContractModel.hs

```diff
@@ -1134,3 +1149,79 @@ checkNoLockedFundsProof options spec NoLockedFundsProof{nlfpMainStrategy   = mai
                               "  " ++ show bal'
                     DL.assert err (bal `leq` bal')
 
+-- | A whitelist entry tells you what final log entry prefixes
+-- are acceptable for a given error
+data WhitelistEntry = WhitelistEntry { acceptEmptyLog :: Bool, errorPrefixes :: [Text.Text] }
+-- | A whitelist associates errors with a whitelist entry
+type Whitelist = Map String WhitelistEntry
+
+-- | Check that a log is accepted by a whitelist entry
+isAcceptedBy :: Maybe Text.Text -> WhitelistEntry -> Bool
+isAcceptedBy Nothing wle          = acceptEmptyLog wle
+isAcceptedBy (Just lastEntry) wle = any (`Text.isPrefixOf` lastEntry) (errorPrefixes wle)
+
+{- Note [Maintaining `whitelistOk` and `checkErrorWhitelist`]
+   The intended use case of `checkErrorWhitelist` is to be able to assert that failures of
+   validation only happen for reasons that the programmer intended. However, to avoid
+   degenerate whitelists that accept obvious mistakes like validation failing because a
+   partial function in the prelude failed we need to make sure that any whitelist used passes
+   the `whitelistOk` check. This means in turn that we need to maintain `whitelistOk` when we
+   introduce new failure modes or change existing failure modes in the prelude or elsewhere
+   in the plutus system.
+-}
+
+{- Note [Divide by zero]
+   Errors that arise from a division by zero fail silently with a `CekEvaluationFailure`. This
```

Comment:
This *shouldn't* be true any more on master, I think. Or at least, the evaluator should give us a better error. Maybe we lose it, hmmm.

Reply:
Aha, I see the issue. It'll come out as an `EvaluationException`. You should look at those too. Those indicate "something went wrong" which amongst other things includes builtins failing.

Reply:
This is I think an error in the doc on the errors...

Reply:
Nope, I'm wrong. The thing we are missing is that the original exception contained a term which caused the error, which the emulator doesn't record. You could use that to detect if it came from a builtin.

Reply:
The cause gets lost here: https://github.com/input-output-hk/plutus/blob/1426912c3f8c29e7ca8754dafae81eaf281e1ee6/plutus-ledger-api/src/Plutus/V1/Ledger/Scripts.hs#L203

Reply:
Here's an example of a div-by-zero error: https://github.com/input-output-hk/plutus/pull/3957/files#diff-d877c33d46c18f0448e116d779fdfb1db237bebce04b87643ef783a3e646cdbeR1

Reply:
It's still a bit annoying as you'll I guess have to just look for an occurrence of "builtin" in there...

Reply:
Right. I dont know how to get around this. Do we have access to the `ErrorWithCause` anywhere "convenient" in the emulator? Do we have enough information at this point to re-run the script evaluation to get at the `ErrorWithCause`? All we really need to know of course is that a builtin failed - any builtin failing should ideally be banned by the `whitelistOk` function.

Reply:
I mean, you can just tweak the line in `Scripts.hs` to print the whole error in that case instead of just the inner bit. Let me make a PR.

Reply:
Ah right, derp. I didn't realise we could touch this part of the code so easily. I guess that makes sense - logs are not under consensus.

Reply:
That function isn't even used by the ledger (it's in a misleading place).

Reply:
Ok great, then I think I have everything I need to fix this. Now, the semantics of Plutus means that when we have a builtin `b` that's failing this is always going to be in a term like `b v1 ... vn`, right?

Reply:
Yeah, I think if you get that kind of failure and the cause is a term with `builtin` in it, then it's a builtin failing.

Reply:
There won't be a log message for a builtin failure, only a particular kind of error.

Reply:
Ok, that's a different approach from what I implemented in e0d4924 - which is to add a log message. We can do it either way.

Reply:
So you're proposing to use something different from the `"CekEvaluationFailure"` string in the case where the failure is due to a builtin - preferrably to use the same string as would be thrown by a "correct" implementation of what's upstream of the code you mentioned?

Reply:
I'm suggesting you just `show` the whole `ErrorWithCause` when it gets shoved into a string. You'd then need your whitelist predicate to accept either the error message or the log message, something like:
- If the error matches X, then accept
- If the error matches Y, then if the last log matches Z, then accept